components/mpas-ocean/src/shared/mpas_ocn_tracer_hmix_redi.F

```diff
@@ -342,32 +394,32 @@ subroutine ocn_tracer_hmix_Redi_tend(meshPool, layerThickness, layerThickEdgeMea
                                 *invAreaCell
                redi_term2_edge(iTr, k, iEdge) = -flux_term2
 
-               dTracerDx = (tracers(iTr, k, cell2) - tracers(iTr, k, cell1))*dvEdge(iEdge)*0.25_RKIND
-               fluxRediZTop(iTr, k) = fluxRediZTop(iTr, k) + &
-                                      0.5_RKIND*(slopeTriadUp(k, iCellSelf, iEdge) + slopeTriadDown(k - 1, iCellSelf, iEdge)) &
-                                      *dTracerDx
+               ! term 3: d/dz div( S grad psi )
+               ! includes weights for edge-to-cell remap
+               fluxRediZTop(iTr, k) = fluxRediZTop(iTr, k) + 0.125_RKIND*dvEdge(iEdge)*( &
+                  slopeTriadDown(k - 1, iCellSelf, iEdge)*(tracers(iTr, k - 1, cell2) - tracers(iTr, k - 1, cell1)) + &
+                  slopeTriadUp(k, iCellSelf, iEdge)*(tracers(iTr, k, cell2) - tracers(iTr, k, cell1)))
             end do
          end do ! nEdgesOnCell(iCell)
 
          if ( use_Redi_diag_terms) then
             ! impose no-flux boundary conditions at top and bottom of column
             fluxRediZTop(:, 1:minLevelCell(iCell)) = 0.0_RKIND
-            fluxRediZTop(:, maxLevelCell(iCell) + 1) = 0.0_RKIND
             redi_term3_topOfCell(:, 1:minLevelCell(iCell), iCell) = 0.0_RKIND
             do k = minLevelCell(iCell) + 1, maxLevelCell(iCell)
                redi_term3_topOfCell(:, k, iCell) = fluxRediZTop(:, k) * invAreaCell
             end do
+            fluxRediZTop(:, maxLevelCell(iCell) + 1) = 0.0_RKIND
             redi_term3_topOfCell(:, maxLevelCell(iCell) + 1, iCell) = 0.0_RKIND
 
+            ! take d/dz and remap div( S grad psi ) from edge to cells
             do k = minLevelCell(iCell), maxLevelCell(iCell)
                do iTr = 1, nTracers
-                  ! Add tendency for Term 3: d/dz ( h S grad phi) = ( S grad phi) fluxes
-                  ! 2.0 in next line is because a dot product on a C-grid
-                  ! requires a factor of 1/2 to average to the cell center.
-                  flux_term3 = rediKappaCell(iCell)*2.0_RKIND* &
-                               (RediKappaScaling(k, iCell)*fluxRediZTop(iTr, k)  &
-                               * invAreaCell - RediKappaScaling(k + 1, iCell) &
-                               *fluxRediZTop(iTr, k + 1) * invAreaCell)
+                  flux_term3 = rediKappaCell(iCell) * ( &
```

Comment:
@dengwirda, the C-grid formula for a dot-product definitely requires a factor of 2.0 to average from edges to cells (I say factor of 1/2 on line 366 above, that should be factor of 2). So here on term 3:

$\partial_{t} h \psi = \dots + \nabla \cdot (\kappa h \nabla \psi) + \nabla \cdot (\kappa h S \partial_{z} \psi) + \partial_{z} (\kappa S \cdot \nabla \psi) + \partial_{z} (\kappa |S|^2 \partial_{z} \psi).$

we are computing $S \cdot \nabla \psi$ on each edge of the cell and interpolating it from edge to cell center. That operation does include a factor of 2. 

The general formula for the dot product at the cell center $(a \cdot b)_c$ based on c-grid normal components $a_e$ and $b_e$ at edges is: 
$(a \cdot b)_c = \frac{2}{N}\sum_{e=1}^N a_e b_e$
(sorry, github is not rendering that one...)

You can see that with a two simple examples. Take two identical vectors $a=(1,0)$ and $b=(1,0)$ in x-y Cartesian coordinates and spatially constant. On a quad c-grid, $a_e b_e$ is 1 on the right and left edges, and zero on the upper and lower edges. Averaging the four edges gives $(1+1+0+0)/4=1/2$, so you need that factor of two to get the correct value of $(a \cdot b)_c = 1$ at the center.

The same example on a hex, with $a=(1,0)$ and $b=(1,0)$, both are have normal values of 1 in the right and left and +/- 1/2 on the other four edges:
![image](https://github.com/E3SM-Ocean-Discussion/E3SM/assets/4036816/037bb233-1894-4226-a01f-37c110bd4023)
Taking the product of normal components $a_e b_e$ at every edge and averaging, we have 
$1/6(2\cdot 1 + 4\cdot 1/4) = 1/6\cdot 3 = 1/2$
so again we need the factor of 2 to get the correct solution of $(a \cdot b)_c = 1$ at the center.

That's not a proof, but it was enough to convince me that the factor of two is required.

Of course, it could be that there is a bug with a factor somewhere else. Next I will [redo the convergence test](https://github.com/MPAS-Dev/MPAS-Model/pull/280#issuecomment-576527278) for every term using exact polynomial solutions, and that will tell us if the code is correct.

Reply:
Thanks @mark-petersen I know what you mean, but that extra 2.0 currently seems to be associated with some strange deformation in the test case. I'll revisit this and double check.